pyflakes/checker.py

```diff
@@ -870,6 +895,63 @@ def ignore(self, node):
     # additional node types
     COMPREHENSION = KEYWORD = FORMATTEDVALUE = handleChildren
 
+    def DICT(self, node):
+        # Complain if there are duplicate keys with different values
+        # If they have the same value it's not going to cause potentially
+        # unexpected behaviour so we'll not complain.
+        keys = [
+            self.convert_to_value(key) for key in node.keys
+        ]
+        values = [
+            self.convert_to_value(value) for value in node.values
+        ]
+        for key in set(keys):
+            if keys.count(key) > 1:
+                candidate_values = []
+                for candidate in range(len(keys)):
+                    if key == keys[candidate]:
+                        candidate_values.append(values[candidate])
+
+                for value in set(candidate_values):
+                    if candidate_values.count(value) == 1:
+                        if isinstance(key, VariableKey):
+                            self.report(messages.DuplicateVariableDictionaryKey,
+                                        node, key.name)
+                        else:
+                            self.report(
+                                messages.DuplicateDictionaryKey,
+                                node,
+                                key,
+                            )
+                        # We only need to alert once per duplicated dictionary
+                        # key
+                        break
+        self.handleChildren(node)
+
+    def convert_to_value(self, item):
+        if isinstance(item, ast.Str):
+            return item.s
+        elif isinstance(item, ast.Tuple):
+            return tuple(self.convert_to_value(i) for i in item.elts)
+        elif isinstance(item, ast.Num):
+            return item.n
+        elif isinstance(item, ast.Name):
+            result = VariableKey(item=item)
+            constants_lookup = {
```

Comment:
I'm sure we will have other uses for this lookup, so it could be a module level constant.